rust-runtime/inlineable/src/endpoint_lib/parse_url.rs

```diff
@@ -0,0 +1,112 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::endpoint_lib::diagnostic::DiagnosticCollector;
+use http::Uri;
+use std::error::Error;
+use url::{Host, Url as ParsedUrl};
+
+#[derive(PartialEq, Debug)]
+pub(crate) struct Url<'a> {
+    uri: Uri,
+    url: ParsedUrl,
+    raw: &'a str,
+}
+
+impl<'a> Url<'a> {
+    pub(crate) fn is_ip(&self) -> bool {
+        match self.url.host() {
+            Some(Host::Ipv4(_) | Host::Ipv6(_)) => true,
+            _ => false,
+        }
+    }
+    pub(crate) fn scheme(&self) -> &str {
+        self.url.scheme()
+    }
+
+    pub(crate) fn authority(&self) -> &str {
+        self.uri.authority().unwrap().as_str()
+    }
+
+    pub(crate) fn normalized_path(&self) -> &str {
+        match self.uri.path() {
+            path if !path.is_empty() => path,
+            _ => "/",
+        }
+    }
+
+    pub(crate) fn path(&self) -> &str {
+        if self.uri.path() == "/" && !self.raw.ends_with('/') {
+            ""
+        } else {
+            self.uri.path()
+        }
+    }
+}
+
+pub(crate) fn parse_url<'a, 'b>(url: &'a str, e: &'b mut DiagnosticCollector) -> Option<Url<'a>> {
+    let raw = url;
+    let uri: Uri = e.capture(url.parse())?;
+    let url: ParsedUrl = e.capture(url.parse())?;
+    if let Some(query) = uri.query() {
+        e.report_error(format!(
+            "URL cannot have a query component (found {})",
+            query
+        ));
+        return None;
+    }
+    if !["http", "https"].contains(&url.scheme()) {
+        e.report_error(format!(
+            "URL scheme must be HTTP or HTTPS (found {})",
+            url.scheme()
+        ));
+        return None;
+    }
+    Some(Url { url, uri, raw })
+}
```

Comment:
It's def overkill but it could be cool to use something like [`miette`](https://docs.rs/miette/latest/miette/) for these error messages